### ISSUE TITLE
how can the input be a dense input in c api

### ISSUE THREAD

**Issue by passionvery** [dmlc/xgboost#3782]
Thanks for participating in the XGBoost community! We use https://discuss.xgboost.ai for any general usage questions and discussions. The issue tracker is used for actionable items such as feature proposals discussion, roadmaps, and bug tracking.  You are always welcomed to post on the forum first :)

Issues that are inactive for a period of time may get closed. We adopt this policy so that we won't lose track of actionable issues that may fall at the bottom of the pile. Feel free to reopen a new one if you feel there is an additional problem that needs attention when an old one gets closed.

For bug reports, to help the developer act on the issues, please include a description of your environment, preferably a minimum script to reproduce the problem.

For feature proposals, list clear, small actionable items so we can track the progress of the change.

**Comment by passionvery** [comment]
Are the two structure the sparse input?(xgboost::SparseBatch  and xgboost::io::DataMatrix)

**Comment by passionvery** [comment]
I train a model in python sklearn using dense input(libsvm data.toarray()).Then I predict the result in python with dense input.The result is different from the prediction in c with both SparseBatch and DataMatrix input.But the prediction in c is the same as the prediction in python with sparse input.So I get the conclusion that the c api inputs (xgboost::SparseBatch and xgboost::io::DataMatrix) are sparse. Is this right?How can I get the dense input in c ? I train the model in python sklearn with both sparse input and dense input.But the dense input has higher accuracy.So I need to know How can I get the dense input in c?Thank you

**Comment by passionvery** [comment]
C:
std::string test_path = "2_output_vaild.fea.com.1";//libsvm
data = xgboost::io::LoadDataMatrix(test_path.c_str(),  silent != 0, use_buffer != 0, loadsplit);
m_learner.Predict(*data, 0, &preds, 0,0);
python:
X_train,Y_train = load_svmlight_file(f_path)
X_train = X_train.toarray()         // The result is the same if this row is deleted,If this row exists,the result is different.
 for x in X_train:
     self.clf._le = LabelEncoder().fit(['label0', 'label1','label2', 'label3'])
      prob_pos = self.clf.predict_proba(x)

**Comment by hcho3** [comment]
The C++ codebase uses sparse format to store data. So dense inputs will need to be converted to sparse (CSR) layout.
